lib/services/local/presence.go

```diff
@@ -678,72 +678,138 @@ func (s *PresenceService) DeleteAllTunnelConnections() error {
+	// unusual way updates are handled in this method meaning that the revision
+	// in the provided remote cluster is not used. We should eventually make a
+	// breaking change to this behavior.
+	const iterationLimit = 3
+	for i := 0; i < iterationLimit; i++ {
+		existing, err := s.GetRemoteCluster(ctx, rc.GetName())
+		if err != nil {
+			return nil, trace.Wrap(err)
+		}
+		existing.SetExpiry(rc.Expiry())
+		existing.SetLastHeartbeat(rc.GetLastHeartbeat())
+		existing.SetConnectionStatus(rc.GetConnectionStatus())
+		existing.SetMetadata(rc.GetMetadata())
 
-	_, err = s.CompareAndSwap(ctx, *existingItem, updateItem)
-	if err != nil {
-		if trace.IsCompareFailed(err) {
-			return trace.CompareFailed("remote cluster %v has been updated by another client, try again", rc.GetName())
+		updateValue, err := services.MarshalRemoteCluster(existing)
+		if err != nil {
+			return nil, trace.Wrap(err)
 		}
-		return trace.Wrap(err)
+
+		lease, err := s.ConditionalUpdate(ctx, backend.Item{
+			Key:      backend.Key(remoteClustersPrefix, existing.GetName()),
+			Value:    updateValue,
+			Expires:  existing.Expiry(),
+			Revision: existing.GetRevision(),
+		})
+		if err != nil {
+			if trace.IsCompareFailed(err) {
+				// Retry!
+				continue
+			}
+			return nil, trace.Wrap(err)
+		}
+		existing.SetRevision(lease.Revision)
+		return existing, nil
+	}
+	return nil, trace.CompareFailed("failed to update remote cluster within %v iterations", iterationLimit)
+}
+
+// GetAndUpdateRemoteCluster fetches a remote cluster and then calls updateFn
+// to apply any changes, before persisting the updated remote cluster.
+func (s *PresenceService) GetAndUpdateRemoteCluster(
+	ctx context.Context,
+	name string,
+	updateFn func(types.RemoteCluster) (types.RemoteCluster, error),
+) (types.RemoteCluster, error) {
+	// Retry to update the remote cluster in case of a conflict.
+	const iterationLimit = 3
+	for i := 0; i < 3; i++ {
+		existing, err := s.GetRemoteCluster(ctx, name)
+		if err != nil {
+			return nil, trace.Wrap(err)
+		}
+
+		updated, err := updateFn(existing.Clone())
+		if err != nil {
+			return nil, trace.Wrap(err)
+		}
+
+		if updated.GetName() != existing.GetName() {
+			return nil, trace.BadParameter("metadata.name: cannot be updated")
+		}
+
+		updatedValue, err := services.MarshalRemoteCluster(updated)
+		if err != nil {
+			return nil, trace.Wrap(err)
+		}
+
+		lease, err := s.ConditionalUpdate(ctx, backend.Item{
+			Key:      backend.Key(remoteClustersPrefix, updated.GetName()),
+			Value:    updatedValue,
+			Expires:  updated.Expiry(),
+			Revision: updated.GetRevision(),
+		})
+		if err != nil {
+			if trace.IsCompareFailed(err) {
+				// Retry!
+				continue
+			}
+			return nil, trace.Wrap(err)
+		}
+		updated.SetRevision(lease.Revision)
+		return updated, nil
 	}
-	return nil
+	return nil, trace.CompareFailed("failed to update remote cluster within %v iterations", iterationLimit)
```

Comment:
Compare failed, or limit exceeded?

Reply:
I noticed some code in the layers above switches behaviour on a CompareFailed - if you feel LimitExceeded is appropriate, I'm happy to change the comparisons at the call sites above. It feels like LimitExceeded doesn't really describe the root cause, and this return can only be hit if it's been a CompareFailed within the loop.